mirgecom/logging_quantities.py

```diff
@@ -0,0 +1,319 @@
+
+    for dim in range(dim):
+        for op in ["min", "max", "norm"]:
+            logmgr.add_quantity(DiscretizationBasedQuantity(
+                discr, "momentum", op, extract_vars_for_logging, units_for_logging,
+                axis=dim))
+
+
+# {{{ Package versions
+
+def add_package_versions(mgr: LogManager, path_to_version_sh: str = None) -> None:
+    """Add the output of the emirge version.sh script to the log.
+
+    Parameters
+    ----------
+    mgr
+        The :class:`logpyle.LogManager` to add the versions to.
+
+    path_to_version_sh
+        Path to emirge's version.sh script. The function will attempt to find this
+        script automatically if this argument is not specified.
+
+    """
+    import subprocess
+    from warnings import warn
+
+    output = None
+
+    # Find emirge's version.sh in any parent directory
+    if path_to_version_sh is None:
+        import pathlib
+        import mirgecom
+
+        p = pathlib.Path(mirgecom.__file__).resolve()
+
+        for d in p.parents:
+            candidate = pathlib.Path(d).joinpath("version.sh")
+            if candidate.is_file():
+                with open(candidate) as f:
+                    if "emirge" in f.read():
+                        path_to_version_sh = str(candidate)
+                        break
+
+    if path_to_version_sh is None:
+        warn("Could not find emirge's version.sh.")
+
+    else:
+        try:
+            output = subprocess.check_output(path_to_version_sh)
+        except OSError as e:
+            warn("Could not record emirge's package versions: " + str(e))
+
+    mgr.set_constant("emirge_package_versions", output)
+
+# }}}
+
+
+# {{{ State handling
+
+def set_sim_state(mgr: LogManager, dim, state, eos) -> None:
+    """Update the simulation state of all :class:`StateConsumer` of the log manager.
+
+    Parameters
+    ----------
+    mgr
+        The :class:`logpyle.LogManager` to set the state of.
+    """
+    state_vars = {}
+
+    for gd_lst in [mgr.before_gather_descriptors,
+            mgr.after_gather_descriptors]:
+        for gd in gd_lst:
+            if isinstance(gd.quantity, StateConsumer):
+                extract_state_vars_func = gd.quantity.extract_state_vars
+                if extract_state_vars_func not in state_vars:
+                    state_vars[extract_state_vars_func] = \
+                        extract_state_vars_func(dim, state, eos)
+
+                gd.quantity.set_state_vars(state_vars[extract_state_vars_func])
+
+
+class StateConsumer:
+    """Base class for quantities that require a state for logging.
+
+    .. automethod:: __init__
+    .. automethod:: set_state_vars
+    """
+
+    def __init__(self, extract_vars_for_logging: Callable):
+        """Store the function to extract state variables.
+
+        Parameters
+        ----------
+        extract_vars_for_logging
+            Returns a dict of the state vars for a particular state.
```

Comment:
What are the arguments? What are the types of key and value of that dict?

Reply:
Added